oak_services/src/http/mod.rs

```diff
@@ -0,0 +1,105 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::proto::oak::encap::{HeaderMap, HeaderValue, HttpRequest, HttpResponse};
+use http::{Request, Response};
+use log::warn;
+use maplit::hashmap;
+use oak_abi::OakStatus;
+use std::convert::TryInto;
+
+impl From<Response<Vec<u8>>> for HttpResponse {
+    fn from(response: Response<Vec<u8>>) -> Self {
+        HttpResponse {
+            body: response.body().to_owned(),
+            status: response.status().as_u16() as i32,
+            headers: Some(HeaderMap::from(response.headers().to_owned())),
+        }
+    }
+}
+
+impl TryInto<Request<Vec<u8>>> for HttpRequest {
+    type Error = OakStatus;
+
+    fn try_into(self) -> Result<Request<Vec<u8>>, Self::Error> {
+        let mut builder = Request::builder()
+            .method(self.method.as_bytes())
+            .uri(self.uri);
+        if let Some(headers) = self.headers {
+            for (header_name, header_value) in headers.to_vec() {
+                builder = builder.header(header_name, header_value);
+            }
+        }
+        builder.body(self.body).map_err(|err| {
+            warn!("Could not create body from vector: {}", err);
+            OakStatus::ErrInternal
+        })
+    }
+}
+
+impl From<http::header::HeaderMap<http::header::HeaderValue>> for HeaderMap {
+    fn from(http_headers: http::header::HeaderMap<http::header::HeaderValue>) -> Self {
+        #![allow(clippy::mutable_key_type)]
+        let mut headers = hashmap! {};
+        for (header_name, header_value) in &http_headers {
+            headers
+                .entry(header_name)
+                .or_insert_with(Vec::new)
+                .push(header_value.as_bytes().to_vec())
+        }
+        let headers = headers
+            .iter()
+            .map(|(name, value)| {
+                (
+                    name.to_string(),
+                    HeaderValue {
+                        values: value.to_owned(),
+                    },
+                )
+            })
+            .collect();
+
+        HeaderMap { headers }
+    }
+}
+
+impl HeaderMap {
+    // Flatten the headers into a vector of tuples (String, http::header::HeaderValue) that can
+    // be easily iterated over when populating the headers in the hyper Response builder.
+    pub fn to_vec(&self) -> Vec<(http::header::HeaderName, http::header::HeaderValue)> {
```

Comment:
I'm guessing this may as well return an `impl Iterator`, and avoid the `collect` at the end, so that everything is still lazily evaluated and may be optimized based on what the caller actually needs?